contracts/strategies/AbstractStrategy.sol

```diff
@@ -0,0 +1,148 @@
+    using SafeERC20 for IERC20;
+    using Address for address;
+
+    uint256 constant MAX_INT = 2**256 - 1;
+    uint256 constant PRICE_DECIMALS = 1e8;
+
+    address public controller;
+    address public supplyToken;
+    address public lpToken;
+
+    uint256 public shares;
+
+    constructor(
+        address _controller,
+        address _supplyToken,
+        address _lpToken
+    ) {
+        controller = _controller;
+        supplyToken = _supplyToken;
+        lpToken = _lpToken;
+    }
+
+    modifier onlyController() {
+        require(msg.sender == controller, "caller is not controller");
+        _;
+    }
+
+    modifier onlyOwnerOrController() {
+        require(msg.sender == owner() || msg.sender == controller, "caller is not owner or controller");
+        _;
+    }
+
+    /**
+     * @return The underlying asset amount of the supply token with 18 decimals
+     */
+    function getAssetAmount() public view virtual returns (uint256);
+
+    /**
+     * @notice Buys lp token from defi contract
+     * @param _buyAmount the amount of underlying asset to be deposited
+     * @return The obtained amount of underlying asset after execution
+     */
+    function buy(uint256 _buyAmount) internal virtual returns (uint256);
+
+    /**
+     * @notice Sells lp token to defi contract for supply token
+     * @param _sellAmount the amount of supply token intended to be redeemed from defi
+     * @return The amount of supply tokens redeemed
+     */
+    function sell(uint256 _sellAmount) internal virtual returns (uint256);
+
+    function getAssetAddress() external view override returns (address) {
+        return lpToken;
+    }
+
+    function aggregateOrders(
+        uint256 _buyAmount,
+        uint256 _sellShares,
+        uint256 _minSharesFromBuy,
+        uint256 _minAmountFromSell
+    ) external override onlyController returns (uint256, uint256) {
+        require(msg.sender == controller, "Not controller");
+        require(shares >= _sellShares, "not enough shares to sell");
+
+        uint256 amountFromSell;
+        uint256 sharesFromBuy;
+        uint256 sharePrice = this.syncPrice();
+
+        if (shares == 0) {
+            shares = _buyAmount;
+            sharesFromBuy = _buyAmount;
+        } else {
+            amountFromSell = (_sellShares * sharePrice) / PRICE_DECIMALS;
+            sharesFromBuy = (_buyAmount * PRICE_DECIMALS) / sharePrice;
+        }
+
+        if (amountFromSell < _buyAmount) {
+            uint256 buyAmount = _buyAmount - amountFromSell;
+            uint256 actualSharesFromBuy = _doBuy(buyAmount);
+            shares += actualSharesFromBuy;
+            uint256 totalSharesFromBuy = actualSharesFromBuy + _sellShares;
+            require(totalSharesFromBuy >= _minSharesFromBuy, "failed min shares from buy");
+            emit Buy(_buyAmount, totalSharesFromBuy);
+            emit Sell(_sellShares, amountFromSell);
+            return (totalSharesFromBuy, amountFromSell);
```

Comment:
redundant code, record values e.g., (`sharesFromBuy = actualSharesFromBuy + _sellShares`), then emit and return only once at the end of the function.

Reply:
wouldn't that change the idiom of sharesFromBuy? or I could elevate the scope of totalSharesFromBuy and totalAmountFromSell.